### PR TITLE
feat(.net): embed package icon when configured

### PR DESCRIPTION
**Pull request by RomainMuller**
When an `iconUrl` is configured for a .NET target, attempt to download
it for inclusion in the NuGet package with the `PackageIcon` attribute,
as the `PackageIconUrl` attribute is deprecated.

This feature can be opted out of by setting the
`JSII_PACMAK_DOTNET_NO_DOWNLOAD_ICON` environment variable (no matter
what value it has).

If the download somehow fails, the previous behavior is preserved.

The `PackageIconUrl` attribute is still emitted for backwards
compatibility with tools that do not (yet) support `PackageIcon`.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts [hunk 1]
```diff
@@ -103,6 +106,27 @@ export class DotNetGenerator extends Generator {
     await fs.mkdirp(path.join(outdir, packageId));
     await fs.copyFile(tarball, path.join(outdir, packageId, tarballFileName));
 
+    // Attempt to download the package icon from the configured URL so we can use the non-deprecated PackageIcon
+    // attribute. If this fails or is opted out (via $JSII_PACMAK_DOTNET_NO_DOWNLOAD_ICON being set), then only the
+    // deprecated PackageIconUrl will be emitted.
+    const iconFile =
```
Comment (comcalvi) [inline packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts:112]: Is there a way for the user to pass this file without downloading it (eg if they already have the icon locally)?

**Review by comcalvi — COMMENTED** [review]
Just a few comments / questions

File: packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts [hunk 2]
```diff
@@ -1191,3 +1215,108 @@ export class DotNetGenerator extends Generator {
...
+    case 'http:':
+      request = http.get;
+      break;
+    case 'https:':
+      request = https.get;
+      break;
+    default:
+      // Unhandled protocol... ignoring
+      debug(
+        `Unsupported URL protocol for resource download: ${url.protocol} (full URL: ${urlText})`,
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts:1236]: Maybe... but then what are the chances this forgets to be updated when more protocols are supported? This won't fail the build, too... it's just a debug message.

File: packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts [hunk 3]
```diff
@@ -1191,3 +1215,108 @@ export class DotNetGenerator extends Generator {
...
+                fileName = `${fileName}.jpg`;
+              }
+              break;
+            case 'image/png':
+              if (!fileName.endsWith('.png')) {
+                fileName = `${fileName}.png`;
+              }
+              break;
+            default:
+            // Nothing to do...
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts:1264]: The spec unfortunately does not specify what is a valid MIME type for the icon, and there are more image formats that I can't be bothered with (ico, heic, etc..)

File: packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts (hunk 1 above)
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts:112]: Not currently... That'd require they bundle the icon in their npm package, which I'm not sure is appropriate...

File: packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts [hunk 4]
```diff
@@ -103,6 +106,27 @@ export class DotNetGenerator extends Generator {
...
+      this.assembly.targets?.dotnet?.iconUrl != null &&
+      !process.env.JSII_PACMAK_DOTNET_NO_DOWNLOAD_ICON
+        ? await tryDownloadResource(
+            this.assembly.targets.dotnet.iconUrl,
+            path.join(outdir, packageId),
+          ).catch((err: any) => {
+            debug(
+              `[dotnet] Unable to download package icon, will only use deprecated PackageIconUrl attribute: ${err.cause}`,
+            );
+            return Promise.resolve(undefined);
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts:122]: I guess it'd do the same, yeah... I kind of use this form to make it clear I turned the rejecting promise into a resolving promise at this point.

**Review by comcalvi — APPROVED** [review]
LGTM!

**Pull request merged**